Updates Haskell LTS version to 18.28

gijsvcuyck opened the pull request:
Updates the LTS version used by the project, allowing the use of newer libraries, and the [haskell-language-server implementation of the LSP](https://github.com/haskell/haskell-language-server) for IDE support while developing.

gijsvcuyck committed:
Fixes several build issues in the ui module.

The Network module is removed, and has been replaced with a combination of the Network.Socket module, and functions from TextViaSockets.
Exception handling is removed from the haskeline library. It is replaced with the exceptions library instead.

gijsvcuyck committed:
Fixes the valexpr module

The monoid typeclass has split into Monoid and Semigroup, making old implementations of Monoid incorrect.
This is fixed by moving the implementation of mappend to the implementation of <> in a new instance of semigroup

gijsvcuyck committed:
Remove redundant import Data.Monoid statements

gijsvcuyck committed:
Fixes serverenv and core modules

Mostly removes redundant monad imports, and replaces the Network library with Network.Socket.

gijsvcuyck committed:
Fixes server module

Replaces the removed Network module with parts from Network.Socket, and parts from TextOverSockets.

Definately needs testing.

gijsvcuyck committed:
Fixes test files

Add Semigroup instance, and replace some calls to Network module with calls to Network.Socket module.

gijsvcuyck committed:
fixed minor typo in build file that got stuck in cached results and broke everything

gijsvcuyck committed:
Updated integration test ghc version to be the same as main project ghc version

gijsvcuyck committed:
Sets all resolver versions spread out over the project to the same resolver, and adds notes to each file pointing to the need to keep the other files updated if things change again in the future.

gijsvcuyck committed:
Fixes integration tests building

tests build now, but they do not execute because they expect the torxakis executable to be on the path, wich messes with global instalations.

gijsvcuyck committed:
Start of fixing benchmark sanity tests

Try to standardize the working directory for all tests to the main git folder. Some tests were running in the sqatt folder, while others were not.
This makes using relative paths to the torxakis executable impossible.

Currently the benchmark sanity tests are still failing nondeterministically.

gijsvcuyck committed:
Remove parallel test execution

parallel test execution was giving nondeterministic errors.
Errors seemed to not be related to the body of the tests, as replacing all tests with 1 `shouldbe` 10 still crashed.
This only seemed to happen when at least 3 testsets where given as an argument to the testExampleSets function, but more frequently with more testsets being used.

observed error:
$ stack test sqatt:benchmarks-sanity-test
sqatt> test (suite: benchmarks-sanity-test)

sqatt> Test suite benchmarks-sanity-test failed

Error: [S-7282]
       Stack failed to execute the build plan.

       While executing the build plan, Stack encountered the error:

       Error: [S-1995]
       Test suite failure for package sqatt-0.1
           benchmarks-sanity-test:  exited with: ExitFailure (-1073741675)

expected behaviour: something like
$ stack test sqatt:benchmarks-sanity-test
sqatt> test (suite: benchmarks-sanity-test)

Debug set 1
  dummy test. FAILED [1]
Debug set 1
  dummy test. FAILED [2]
Debug set 1
  dummy test. FAILED [3]

Failures:

  src\Sqatt.hs:497:46:
  1) Debug set 1 dummy test.
       expected: 10
        but got: 1

  To rerun use: --match "/Debug set 1/dummy test./"

  src\Sqatt.hs:497:46:
  2) Debug set 1 dummy test.
       expected: 10
        but got: 1

  To rerun use: --match "/Debug set 1/dummy test./"

  src\Sqatt.hs:497:46:
  3) Debug set 1 dummy test.
       expected: 10
        but got: 1

  To rerun use: --match "/Debug set 1/dummy test./"

Randomized with seed 1632928150

Finished in 0.0006 seconds
3 examples, 3 failures

gijsvcuyck committed:
Remove more parallel test execution, and support for czc4

gijsvcuyck committed:
Fixes integration tests.

gijsvcuyck committed:
Fixes benchmarks, and updates testing readme

gijsvcuyck committed:
Disabled movingarms test in exploremodels testset.

Test removed as it does not terminate, tested in latest develop, v0.9 and v0.6,

gijsvcuyck committed:
General fixes to become compatible with newer version of ghc.

Mostly removed redundant imports and error cases.
Temporarily adds Unique as extra dependency, because replacing the depriciated package is probably not worth it if the module in which it is used (txs-compiler) is planned to be phased out.

gijsvcuyck committed:
Fixes unit tests for newer version of ghc

gijsvcuyck committed:
Fixes integration of changes into system tests.

gijsvcuyck committed:
Changed dynamic testset to use same pathing convention as other tests

Default is relative path starting in root git directory.
Relative paths starting in sqatt directory need to use explicit insqatt modifier.

gijsvcuyck committed:
Merge pull request #1 from gijsvcuyck/rebaseCompilerVersionUpdate

Update LTS version for newer ghc compiler

keremispirli commented on stack.yaml:
```diff
@@ -31,15 +31,17 @@ packages:
 extra-deps:
-- git: https://github.com/TorXakis/text-via-sockets.git
-  commit: e3228cd0407ec0d7991a544e154ea2def184fcae
+- git: https://github.com/gijsvcuyck/text-via-sockets.git
```
Why not merge this fork back first? Then you won't need to change this reference.

gijsvcuyck replied:
> Why not merge this fork back first? Then you won't need to change this reference.

I just made both pull requests at the same time, since I figured it might take a while before anything got merged. The other pull request is [here](https://github.com/TorXakis/text-via-sockets/pull/5).
It would indeed be better to merge the the test-via-sockets changes first, but it also would not be that hard to change the reference again later.

pjljvandelaar commented on README.md:
````diff
@@ -83,6 +83,8 @@ stack setup --stack-yaml stack_linux.yaml
 stack install --stack-yaml stack_linux.yaml
 ```
 
+```stack install``` makes the resulting executable availiable globaly. For developing purposes, ```stack build``` can be used instead. The resulting executable can then be accessed using ```stack exec torxakis -- <torxakis_options> ```.
````
typo: available

pjljvandelaar commented on stack.yaml:
```diff
@@ -31,15 +31,17 @@ packages:
   hexpat:
     bundle: true
-  text:
-    integer-simple: true
```
Indeed integer-simple is no longer needed as Haskell is change to use it.

gijsvcuyck replied:
When I have time, I will try to remove all mentions of integer-simple and rerun the tests.

gijsvcuyck replied:
I did some googling, and the available information is rather vague again, but it looks like the change in haskell to move away from integer-simple happened around ghc 9, and this pr moves the ghc version up to ghc 8.10.7.

I also heard that integer-simple was originally used for licensing reasons, so I cannot access if it is ok to move away from it myself. As this seems a separate issue from the rest of the pr, I don't think it should be changed here. 

The above flag change is separate from this as the flag itself no longer exists. I could not find any documentation on when or why this changed or if it needed to be replaced with something, but removing it did not break any of the tests at least.

pjljvandelaar commented on sys/lpeops/src/core/LPEContextIds.hs:
```diff
@@ -127,7 +127,6 @@ getValExprIds = customData . visitValExpr searchVisitor
                     (view -> Vconcat _)               -> idsInSubExps
                     (view -> Vstrinre _ _)            -> idsInSubExps
                     (view -> Vpredef _ fid _)         -> Set.insert (TxsDefs.IdFunc fid) idsInSubExps
-                    _                                 -> error ("GetValExprIds.searchVisitor not defined for " ++ show expr ++ "!")
```
Why did you remove these error messages to the user?
LPE doesn't support all construct (e.g. not dynamic process creation),
user should be informed when he is using LPE incorrectly.

gijsvcuyck replied:
I removed the default case here, and in several other places, because ghc was giving warnings about duplicate cases. According to ghc, all possible cases of expr were already covered, and so the default case was unnecessary. Since warnings are treated as errors, I had to remove the default case to get it to compile.

I double checked, and the case distinction here is over a value of type `ValExprView VarId`, and all possible constructors of this type (TorXakis\sys\valexpr\src\ValExprDefs.hs line51) are indeed being covered here.

pjljvandelaar commented on test/environment/stack.yaml:
```diff
@@ -10,16 +10,10 @@
+
+# NOTE: This has to manually be kept up to date with the resolver used in the main stack.yaml file.
+resolver: lts-18.28
 ghc-variant: integersimple
```
Don't think this ghc-variant is needed anymore.

pjljvandelaar commented on test/sqatt/src/Benchmarks/Sequence.hs:
```diff
@@ -52,28 +52,29 @@ sequence10Ints = emptyExample
-    , txsServerArgs = ["--smt-solver", "cvc4"]
-      , expectedResult = Pass
-    }
+-- cvc4 not currently supported.
```
You could use cvc5 instead

gijsvcuyck replied:
I tried, but the interface changed, and I did not want to put time into fixing features that were already broken and I am not going to use. See #961. This would be something for a separate pull request anyway I think.

pjljvandelaar commented on test/sqatt/stack.yaml:
```diff
@@ -10,10 +10,11 @@
-resolver: lts-11.22
+# Note: This has to manually be kept up to date with the resolver version in the main stack.yaml file, or the test results might not match with the release build.
+resolver: lts-18.28
 ghc-variant: integersimple
```
ghc-variant can be removed

gijsvcuyck committed:
Fixes typo

Commit:
Fix linux stack.yaml to be in line with windows stack.yaml

pjljvandelaar reviewed (dismissed):
Indeed don't address all problems at once!
I assume for all not addressed problems new issues are made (and linked to this review).

gijsvcuyck committed:
Removes integersimple ghc variant

All the tests still pass without it, and the original licensing issue seems to have been resolved.

pjljvandelaar commented on test/environment/stack.yaml:
```diff
@@ -41,9 +39,9 @@ packages:
 flags: 
   hexpat:
+    # Set for https://github.com/TorXakis/TorXakis/issues/796.
+    # However, this seems like a linux oriented option so it is a bit strange to have this is the windows stack file.
```
typo 
to have this i*n* the windows stack file

pjljvandelaar commented on stack.yaml:
```diff
@@ -41,13 +40,9 @@ extra-deps:
 flags:
   hexpat:
+    # Set for https://github.com/TorXakis/TorXakis/issues/796. 
+    # However, this seems like a linux oriented option so it is a bit strange to have this is the windows stack file.
```
typo
to have this i*n* the windows stack file

gijsvcuyck committed:
Updates version of hspec used

Fixes #978
